File: booking-service/app/repositories/booking_repository.py
import psycopg2.extras
from app.config.database import get_db


def _dict(cursor_row) -> dict:
    return dict(cursor_row)


def create_booking(user_id: str, event_id: str, quantity: int) -> dict:
    with get_db() as conn:
        with conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor) as cur:
            cur.execute(
                "INSERT INTO bookings (user_id, event_id, quantity, status) "
                "VALUES (%s, %s, %s, 'pending') RETURNING *",
                (user_id, event_id, quantity),
            )
            row = cur.fetchone()
            conn.commit()
            return _dict(row)


def get_booking_by_id(booking_id: str) -> dict | None:
    with get_db() as conn:
        with conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor) as cur:
            cur.execute("SELECT * FROM bookings WHERE id = %s", (booking_id,))
            row = cur.fetchone()
            return _dict(row) if row else None


def get_bookings_by_user(user_id: str) -> list[dict]:
    with get_db() as conn:
        with conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor) as cur:
            cur.execute(
                "SELECT * FROM bookings WHERE user_id = %s ORDER BY created_at DESC",
                (user_id,),
            )
            return [_dict(r) for r in cur.fetchall()]


def update_booking_status(booking_id: str, status: str) -> dict | None:
    with get_db() as conn:
        with conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor) as cur:
            cur.execute(
                "UPDATE bookings SET status = %s, updated_at = NOW() WHERE id = %s RETURNING *",
                (status, booking_id),
            )
            row = cur.fetchone()
            conn.commit()
            return _dict(row) if row else None


def get_event_capacity(event_id: str) -> dict | None:
    with get_db() as conn:
        with conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor) as cur:
            cur.execute("SELECT * FROM event_capacity WHERE event_id = %s", (event_id,))
            row = cur.fetchone()
            return _dict(row) if row else None


def reserve_seats_optimistic(event_id: str, quantity: int, total_seats: int) -> dict:
    """
    Atomically reserve seats using optimistic locking on the version column.
    Returns:
        {"success": True, "capacity": {...}}
        {"success": False, "reason": "insufficient_seats", "available": int}
        {"success": False, "reason": "concurrent_modification"}
    """
    with get_db() as conn:
        conn.autocommit = False
        try:
            with conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor) as cur:
                cur.execute("SELECT * FROM event_capacity WHERE event_id = %s", (event_id,))
                capacity = cur.fetchone()

                if capacity is None:
                    cur.execute(
                        "INSERT INTO event_capacity (event_id, total_seats, reserved_seats, version) "
                        "VALUES (%s, %s, 0, 0) "
                        "ON CONFLICT (event_id) DO UPDATE SET total_seats = EXCLUDED.total_seats "
                        "RETURNING *",
                        (event_id, total_seats),
                    )
                    capacity = cur.fetchone()

                capacity = _dict(capacity)
                available = capacity["total_seats"] - capacity["reserved_seats"]

                if available < quantity:
                    conn.rollback()
                    return {"success": False, "reason": "insufficient_seats", "available": available}

                # Optimistic lock: only update if version is unchanged since we read it
                cur.execute(
                    "UPDATE event_capacity "
                    "SET reserved_seats = reserved_seats + %s, version = version + 1 "
                    "WHERE event_id = %s AND version = %s AND (total_seats - reserved_seats) >= %s "
                    "RETURNING *",
                    (quantity, event_id, capacity["version"], quantity),
                )
                updated = cur.fetchone()

                if updated is None:
                    conn.rollback()
                    return {"success": False, "reason": "concurrent_modification"}

                conn.commit()
                return {"success": True, "capacity": _dict(updated)}
        except Exception:
            conn.rollback()
            raise
        finally:
            conn.autocommit = True


def release_seats(event_id: str, quantity: int) -> None:
    with get_db() as conn:
        with conn.cursor() as cur:
            cur.execute(
                "UPDATE event_capacity "
                "SET reserved_seats = GREATEST(0, reserved_seats - %s), version = version + 1 "
                "WHERE event_id = %s",
                (quantity, event_id),
            )
            conn.commit()

File: booking-service/src/repositories/bookingRepository.js
const { pool } = require('../config/db');

async function createBooking(userId, eventId, quantity) {
  const result = await pool.query(
    `INSERT INTO bookings (user_id, event_id, quantity, status)
     VALUES ($1, $2, $3, 'pending') RETURNING *`,
    [userId, eventId, quantity]
  );
  return result.rows[0];
}

async function updateBookingStatus(bookingId, status) {
  const result = await pool.query(
    `UPDATE bookings SET status = $1, updated_at = NOW() WHERE id = $2 RETURNING *`,
    [status, bookingId]
  );
  return result.rows[0] || null;
}

async function getBookingById(id) {
  const result = await pool.query('SELECT * FROM bookings WHERE id = $1', [id]);
  return result.rows[0] || null;
}

async function getBookingsByUser(userId) {
  const result = await pool.query(
    'SELECT * FROM bookings WHERE user_id = $1 ORDER BY created_at DESC',
    [userId]
  );
  return result.rows;
}

async function getEventCapacity(eventId) {
  const result = await pool.query(
    'SELECT * FROM event_capacity WHERE event_id = $1',
    [eventId]
  );
  return result.rows[0] || null;
}

async function _getOrCreateCapacity(client, eventId, totalSeats) {
  const existing = await client.query(
    'SELECT * FROM event_capacity WHERE event_id = $1',
    [eventId]
  );
  if (existing.rows.length > 0) return existing.rows[0];

  const created = await client.query(
    `INSERT INTO event_capacity (event_id, total_seats, reserved_seats, version)
     VALUES ($1, $2, 0, 0)
     ON CONFLICT (event_id) DO UPDATE SET total_seats = EXCLUDED.total_seats
     RETURNING *`,
    [eventId, totalSeats]
  );
  return created.rows[0];
}

// Optimistic locking: atomically reserves seats.
// Returns { success: true, capacity } or { success: false, reason }
async function reserveSeatsOptimistic(eventId, quantity, totalSeats) {
  const client = await pool.connect();
  try {
    await client.query('BEGIN');

    const capacity = await _getOrCreateCapacity(client, eventId, totalSeats);
    const available = capacity.total_seats - capacity.reserved_seats;

    if (available < quantity) {
      await client.query('ROLLBACK');
      return { success: false, reason: 'insufficient_seats', available };
    }

    // Only succeeds if version hasn't changed since we read it (optimistic lock)
    const updated = await client.query(
      `UPDATE event_capacity
       SET reserved_seats = reserved_seats + $1, version = version + 1
       WHERE event_id = $2 AND version = $3 AND (total_seats - reserved_seats) >= $1
       RETURNING *`,
      [quantity, eventId, capacity.version]
    );

    if (updated.rowCount === 0) {
      await client.query('ROLLBACK');
      return { success: false, reason: 'concurrent_modification' };
    }

    await client.query('COMMIT');
    return { success: true, capacity: updated.rows[0] };
  } catch (err) {
    await client.query('ROLLBACK');
    throw err;
  } finally {
    client.release();
  }
}

async function releaseSeats(eventId, quantity) {
  await pool.query(
    `UPDATE event_capacity
     SET reserved_seats = GREATEST(0, reserved_seats - $1), version = version + 1
     WHERE event_id = $2`,
    [quantity, eventId]
  );
}

module.exports = {
  createBooking,
  updateBookingStatus,
  getBookingById,
  getBookingsByUser,
  getEventCapacity,
  reserveSeatsOptimistic,
  releaseSeats,
};
